core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/util/KeyStoreManager.java

```diff
@@ -260,6 +312,165 @@ public KeyStore getKeyStore(String keyStoreName) throws Exception {
+        if (!isSuperTenant) {
+            org.wso2.carbon.registry.api.Association[] associations = registry.getAssociations(
+                    keyStoreFullname, RegistryResources.SecurityManagement.ASSOCIATION_TENANT_KS_PUB_KEY);
+
+            if (associations != null && associations.length > 0) {
+                Resource pubKeyResource = registry.get(associations[0].getDestinationPath());
+                String fileName = generatePubCertFileName(keyStoreFullname, pubKeyResource.getProperty(
+                        RegistryResources.SecurityManagement.PROP_TENANT_PUB_KEY_FILE_NAME_APPENDER));
+
+                if (MessageContext.getCurrentMessageContext() != null) {
+                    String pubKeyFilePath =
+                            dumpCert(MessageContext.getCurrentMessageContext().getConfigurationContext(),
+                                    (byte[]) pubKeyResource.getContent(), fileName);
+                    keyStoreMetadata.setPubKeyFilePath(pubKeyFilePath);
+                }
+            }
+        }
+        return keyStoreMetadata;
+    }
+
+    private KeyStoreMetadata getPrimaryKeyStoreMetadata() {
+
+        ServerConfiguration config = ServerConfiguration.getInstance();
+        String fileName = config.getFirstProperty(RegistryResources.SecurityManagement.SERVER_PRIMARY_KEYSTORE_FILE);
+        String type = config.getFirstProperty(RegistryResources.SecurityManagement.SERVER_PRIMARY_KEYSTORE_TYPE);
+        String name = KeyStoreUtil.getKeyStoreFileName(fileName);
+
+        KeyStoreMetadata primaryKeyStoreMetadata = new KeyStoreMetadata();
+        primaryKeyStoreMetadata.setKeyStoreName(name);
+        primaryKeyStoreMetadata.setKeyStoreType(type);
+        primaryKeyStoreMetadata.setProvider(" ");
+        primaryKeyStoreMetadata.setPrivateStore(true);
+        return primaryKeyStoreMetadata;
+    }
+
+    /**
+     * This method is used to generate the file name of the public cert of a tenant.
+     *
+     * @param ksLocation Keystore location in the registry.
+     * @param uuid       UUID appender.
+     * @return file name of the public cert.
+     */
+    private String generatePubCertFileName(String ksLocation, String uuid) {
+
+        String tenantName = ksLocation.substring(ksLocation.lastIndexOf("/"));
+        for (KeystoreUtils.StoreFileType fileType : KeystoreUtils.StoreFileType.values()) {
+            String fileExtension = KeystoreUtils.StoreFileType.getExtension(fileType);
+            if (tenantName.endsWith(fileExtension)) {
+                tenantName = tenantName.replace(fileExtension, "");
+            }
+        }
+        return tenantName + "-" + uuid + ".cert";
+    }
+
+    /**
+     * Get keystore type.
+     *
+     * @param keyStoreName Keystore name.
+     * @return Keystore type.
+     * @throws SecurityException If an error occurs while retrieving the keystore type.
+     */
+    public String getKeyStoreType(String keyStoreName) throws SecurityException {
```

Comment:
Since we can get the type from the `KeyStoreMetadata`, `KeyStore` classes, check whether this method is really needed.

Reply:
Addressed with https://github.com/wso2/carbon-kernel/pull/4099/commits/f89bbb95d862ca9f626a21592685bbd3e230eb11